src/Ast.ml

```diff
@@ -335,6 +335,113 @@ module Signature_item : Module_item with type t = signature_item = struct
+    | Pexp_field (e, _) -> expr_has_doc e
+    | Pexp_setfield (e1, _, e2) -> expr_has_doc e1 || expr_has_doc e2
+    | Pexp_array el -> List.exists el ~f:expr_has_doc
+    | Pexp_ifthenelse (e1, e2, eo) ->
+        expr_has_doc e1 || expr_has_doc e2
+        || Option.value_map eo ~default:false ~f:expr_has_doc
+    | Pexp_sequence (e1, e2) -> expr_has_doc e1 || expr_has_doc e2
+    | Pexp_while (e1, e2) -> expr_has_doc e1 || expr_has_doc e2
+    | Pexp_for (p, e1, e2, _, e3) ->
+        pattern_has_doc p || expr_has_doc e1 || expr_has_doc e2
+        || expr_has_doc e3
+    | Pexp_constraint (e, ct) -> expr_has_doc e || core_type_has_doc ct
+    | Pexp_coerce (e, cto, ct) ->
+        expr_has_doc e
+        || Option.value_map cto ~default:false ~f:core_type_has_doc
+        || core_type_has_doc ct
+    | Pexp_send (e, _) -> expr_has_doc e
+    | Pexp_new _ -> false
+    | Pexp_setinstvar (_, e) -> expr_has_doc e
+    | Pexp_override _ -> false
+    | Pexp_letmodule (_, me, e) ->
+        expr_has_doc e || not (List.is_empty me.pmod_attributes)
+    | Pexp_letexception (_, e) -> expr_has_doc e
+    | Pexp_assert e -> expr_has_doc e
+    | Pexp_lazy e -> expr_has_doc e
+    | Pexp_poly (e, cto) ->
+        expr_has_doc e
+        || Option.value_map cto ~default:false ~f:core_type_has_doc
+    | Pexp_object cs -> pattern_has_doc cs.pcstr_self
+    | Pexp_newtype (_, e) -> expr_has_doc e
+    | Pexp_pack me -> not (List.is_empty me.pmod_attributes)
+    | Pexp_open (_, _, e) -> expr_has_doc e
+    | Pexp_extension _ -> false
+    | Pexp_unreachable -> false
+
+  and core_type_has_doc ct = not (List.is_empty ct.ptyp_attributes)
+
+  and pattern_has_doc p =
+    pattern_desc_has_doc p.ppat_desc
+    || not (List.is_empty p.ppat_attributes)
+
+  and pattern_desc_has_doc = function
+    | Ppat_any | Ppat_var _ -> false
+    | Ppat_alias (p, _) -> pattern_has_doc p
+    | Ppat_constant _ -> false
+    | Ppat_interval _ -> false
+    | Ppat_tuple lp -> List.exists lp ~f:pattern_has_doc
+    | Ppat_construct (_, op) ->
+        Option.value_map op ~default:false ~f:pattern_has_doc
+    | Ppat_variant (_, op) ->
+        Option.value_map op ~default:false ~f:pattern_has_doc
+    | Ppat_record (lp, _) ->
+        List.exists lp ~f:(fun (_, x) -> pattern_has_doc x)
+    | Ppat_array lp -> List.exists lp ~f:pattern_has_doc
+    | Ppat_or (p1, p2) -> pattern_has_doc p1 || pattern_has_doc p2
+    | Ppat_constraint (p, ct) -> pattern_has_doc p || core_type_has_doc ct
+    | Ppat_type _ -> false
+    | Ppat_lazy p -> pattern_has_doc p
+    | Ppat_unpack _ -> false
+    | Ppat_exception p -> pattern_has_doc p
+    | Ppat_extension _ -> false
+    | Ppat_open (_, p) -> pattern_has_doc p
+
+  let is_simple (i, c) =
+    Poly.(c.Conf.module_item_spacing = `Compact)
+    && Location.width i.pexp_loc <= c.Conf.margin
+    && Location.is_single_line i.pexp_loc
+
+  let break_between (i1, c1) (i2, c2) =
+    expr_has_doc i1 || expr_has_doc i2
```

Comment:
Do you have examples where checking for docs here is necessary? It is not obvious to me that checking `pvb_attributes` is not enough.